examples/CMakeLists.txt

```diff
@@ -12,68 +12,178 @@
+    if ("${PARSED_SKIP_RUN_EXAMPLES}" STREQUAL "")
+        set(PARSED_SKIP_RUN_EXAMPLES "OFF") # Default to OFF.
+    endif ()
+
+    # path/to/source.cpp -> source
+    get_filename_component(name ${source} NAME_WE)
+
+    # path/to/source.cpp -> path/to
+    get_filename_component(subdir ${source} DIRECTORY)
+
+    # path/to/source -> path-to-source
+    string(REPLACE "/" "-" target_name "${subdir}/${name}")
+    add_executable(${target_name} EXCLUDE_FROM_ALL ${source})
+    target_link_libraries(${target_name} PRIVATE ${PARSED_LIBRARIES})
+
+    # Keep executables in separate directories.
+    set_target_properties(${target_name} PROPERTIES
+        OUTPUT_NAME ${name}
+        RUNTIME_OUTPUT_DIRECTORY ${subdir}
+    )
+
+    # Deliberately use `__vectorcall` by default with MSVC to catch missing `__cdecl`.
+    target_compile_options(${target_name} PRIVATE "$<$<CXX_COMPILER_ID:MSVC>:/Gv>")
+
+    # Keep build and run targets completely separate.
+    add_dependencies(examples ${target_name})
+
+    # Use full path to avoid implicit build dependency.
+    if (isMultiConfig)
+        add_custom_target(run-examples-${target_name} COMMAND "${subdir}/$<CONFIG>/${name}")
+    else ()
+        add_custom_target(run-examples-${target_name} COMMAND "${subdir}/${name}")
+    endif ()
+
+    set (type "")
+    if (${source} MATCHES "^api/")
+        set (type "api")
+    elseif (${source} MATCHES "^bsoncxx/")
+        set (type "bsoncxx")
+    elseif (${source} MATCHES "^mongocxx/")
+        set (type "mongocxx")
+    else ()
+        message (FATAL_ERROR "unexpected subdirectory ${subdir}")
+    endif ()
+
+    if (NOT PARSED_SKIP_RUN_EXAMPLES)
+        add_dependencies(run-${type}-examples run-examples-${target_name})
+
+        if (${subdir} MATCHES "^api/")
+            add_dependencies(run-${type}-examples run-examples-${target_name})
+        endif ()
+    endif ()
+endfunction()
+
+file(GLOB_RECURSE api_examples_sources
+    RELATIVE ${CMAKE_CURRENT_SOURCE_DIR}
+    CONFIGURE_DEPENDS
+    "api/*.cpp"
+)
+
+file(GLOB_RECURSE bsoncxx_examples_sources
+    RELATIVE ${CMAKE_CURRENT_SOURCE_DIR}
+    CONFIGURE_DEPENDS
+    "bsoncxx/*.cpp"
+)
+
+file(GLOB_RECURSE mongocxx_examples_sources
+    RELATIVE ${CMAKE_CURRENT_SOURCE_DIR}
+    CONFIGURE_DEPENDS
+    "mongocxx/*.cpp"
+)
+
+foreach(source ${api_examples_sources})
+    add_examples_executable(${source} LIBRARIES mongocxx_target)
+endforeach()
+
+foreach(source ${bsoncxx_examples_sources})
+    add_examples_executable(${source} LIBRARIES bsoncxx_target)
+endforeach()
+
+foreach(source ${mongocxx_examples_sources})
+    set(extra_libs "")
+    set(skip_run_examples OFF)
+
+    if (${source} MATCHES "pool")
+        list(APPEND extra_libs Threads::Threads)
     endif()
 
     # Don't run the CSFLE examples as part of the suite; needs mongocryptd running.
-    if (${EXAMPLE} MATCHES "encryption")
-        continue()
+    if (${source} MATCHES "encryption")
+        set(skip_run_examples ON)
```

Comment:
I find the name `skip_run_examples` quite confusing. On the first read-through I was convinced that anytime a matching example was encountered that it would skip running all examples. I recognize that this variable is distinct from `SKIP_RUN_EXAMPLES` (since CMake is case sensitive when it comes to variable names) and I did eventually figure out that the former is used to set the value of the latter as a property for the individual example executable. However, I recommend renaming this variable to `skip_run_this_example`.

Reply:
Good point. There is some double-negation occurring with the current setup relative to the default:

- Is this example added to run-examples?
  - Yes (by default), unless:
    - No Skip -> Yes (Yes -> No -> Yes) (<--)
    - Yes Skip -> No (Yes -> Yes -> No)

Renamed `SKIP_RUN_EXAMPLES` -> `ADD_TO_RUN_EXAMPLES` and `skip_run_examples` -> `add_to_run_examples` so there is only one level of negation relative to the default.

- Is this example added to run-examples?
  - Yes (by default), unless:
    - No Add -> No (Yes -> No -> No)
    - Yes Add -> Yes (Yes -> Yes -> Yes)

Reply:
Thanks. That is much more clear.